Add a graphical user interface for the vision framework

LeonarddeR opened the pull request:
### Link to issue number:
Follow up of #9064
Closes #971 

### Summary of the issue:
#9064 introduced the vision framework without a graphical user interface.

### Description of how this pull request fixes the issue:
Added a new vision panel to the settings dialog. Vision enhancement providers can provide their own gui panel class to create a customized gui. Otherwise, they can rely on the supportedSettings property to be implemented on the provider.

### Testing performed:
Tested that providers load and unload and that settings can be changed using the gui.

### Known issues with pull request:
* Providers load and unload directly when checking and unchecking check boxes
* [ ] No user guide changes yet. I'd like to agree about the UX before doing that.

### Change log entry:
* New features
    + Added screen highlight functionality to NVDA. (#971, #9064)
        * The NVDA Highlighter is able to highlight the focus, navigator object and browse mode caret position and can be enabled and configured from the new vision category in NVDA's settings dialog.
        * If you have the focus highlight add-on installed, you are encouraged to reconsider its installation. Note that that add-on can still be used as long as the built-in highlighter is disabled.

There's also a wording issue in the changes for developers setting:

* A framework allowing developers to create vision enhancement providers; > please add the word `Àdded` to the start of that sentence. You could also consider updating this entry to point at the new vision category in the gui.

Commit:
Add a graphical user interface for the vision framework

leonardder commented:
@nishimotz I was thinking about your focus highlight add-on while looking at the changelog entry. Are you intending to discontinue your add-on, or will you rebase it onto the new framework? I think there are still some implementation differences that might cause people to prefer one or the other.
Of course, you're also welcome to extend on the current highlighter in core. One thing I'd like to make configurable at some point is the color and thickness of the highlights.

feerrenrut committed:
Layout adjustments to vision framework

feerrenrut reviewed (approved):
What is your plan for the user guide? Could you update it in this PR, covering the vision panel and focus highlight. Then on the screen curtain PR add that?

I hope you don't mind, I have pushed a commit that addresses a few layout issues.

leonardder commented on source/gui/settingsDialogs.py:
```diff
@@ -2858,6 +2866,163 @@ def onBlinkCursorChange(self, evt):
+		)
+		providersSizer = guiHelper.BoxSizerHelper(self, orientation=wx.VERTICAL)
+		providerListDescTextCtrl = providersSizer.addItem(wx.StaticText(self, label=providersListDescriptionText))
+		providerListDescTextCtrl.Wrap(self.scaleSize(544))
```
@feerrenrut: What does 544 mean here?

feerrenrut replied:
This is the width that is available to the text, we actually use the same value in the advanced settings panel. I would really prefer that we use `providerSizer.Size.Width`, except that during construction it is not correct, which I suspect has something to do with the order that we create things / add them to sizers etc. Unfortunately we couldn't use this value just anywhere, it is dependent on the nesting level of the text, and has to be checked manually.

leonardder replied:
I decided to reword your comment here into an explanatory comment block in the text.

leonardder commented on source/gui/settingsDialogs.py:
```diff
@@ -2858,6 +2866,163 @@ def onBlinkCursorChange(self, evt):
+
+		settingsSizerHelper.addItem(providersSizer, flag=wx.EXPAND)
+		settingsSizerHelper.addItem(
+			wx.StaticLine(self, size=wx.Size(-1, self.scaleSize(2))),
```
Also, should -1 and 2 be constants?

feerrenrut replied:
On second thoughts, you can remove the size parameter.

LeonarddeR reviewed (commented):
I'll do the user guide as part of this. Have two questions about your change

Commit:
Update user guide

Typo: vvision > vision

BNrowse mode caret is shown using a rectangle, not a single line

feerrenrut commented:
Overall, I am pretty happy with the default colors for focus highlight. But I might adjust them to have some transparency in a later PR. When they are over text, it can obscure adjacent text.

Commit:
Many gui and framework updates and improvements, including a load warning for the screen curtain

nishimotz commented:
We should consider people who have difficulty distinguishing between red and green colors.
Since the Focus Highlight decided the color first,
I have added the distinction according to the line type later.

Commit:
Fix last (linting) issues

leonardder commented:
@feerrenrut: This is ready for another review. I think everything from 641ccdcc3591e4d3ca6325c0a3bd2f439bf3422c is new.

feerrenrut commented:
I'm seeing some unexpected changes in the diff ie libluis. I'll review the user guide to start with.

feerrenrut commented on user_docs/en/userGuide.t2t:
```diff
@@ -739,6 +744,39 @@ Dot 8 translates any braille input and presses the enter key.
+Furthermore, it might be hard to ensure that there isn't someone looking over your shoulder.
+For this situation, NVDA contains a screen curtain that, when enabled, makes the screen black.
+
+When enabling the Screen Curtain in the [vision category #VisionSettings] of the [NVDA Settings #NVDASettings] dialog, a warning will be displayed telling you that your screen will be black after activating the curtain.
```
```suggestion
Enable the Screen Curtain in the [vision category #VisionSettings] of the [NVDA Settings #NVDASettings] dialog.
A warning that your screen will become black after activation will be displayed.
Before continuing (selecting "Yes"), ensure you have enabled speech / braille and will be able to control your computer without the use of the screen.
Select "No" if you no longer wish to enable the Screen Curtain.
```

dingpengyu commented:
Current pr build failed
Can it be fixed?
thank

Commit:
Revert accidental change to liblouis submodule

leonardder commented:
@dingpengyu Failed appveyor builds are pretty likely to be acted upon pretty soon after they fail.

Commit:
Fix translator comments

LeonarddeR committed:
Apply user guide suggestions from code review

Co-Authored-By: Reef Turner <feerrenrut@users.noreply.github.com>

Commit:
Apply additional review action for the highlighter user guide description

dingpengyu commented:
> @dingpengyu Failed appveyor builds are pretty likely to be acted upon pretty soon after they fail.
problem solved
thank

JulienCochuyt commented:
I was afraid that the Highlighter could suffer from #10211, that is not being functional while the Windows Magnifier is turned on.
It is not.
Thanks to the GUI implemented in this PR, I could successfully test the Highlighter with Windows Magnifier on and everything seems to work as expected.

leonardder commented:
Note that this gui is still a prototype, it will have a revisit by 
@feerrenrut

feerrenrut committed:
Tidy ctypes declarations for screen curtain

feerrenrut committed:
List each vision provider in it's own group.

Distinguish between runtime and static settings for providers.

feerrenrut committed:
Use a custom GUI for highlighter

feerrenrut committed:
Highligher custom gui works

feerrenrut committed:
Highlighter works without custom gui

feerrenrut committed:
dynamic addition of settings at runtime is possible with auto (not custom) settings GUI

feerrenrut committed:
move driver settings and utils

feerrenrut committed:
Move percent/param conversion helpers

feerrenrut committed:
Move settings related content to AutoSettings class

feerrenrut committed:
Tidy up config save reg / unreg

feerrenrut committed:
Adapt identifier, section, and productName to AutoSettings class

These can now be overriden, Driver maps them to name, section
and description.

feerrenrut committed:
Add type hinting and comments

feerrenrut committed:
Fix: don't hide 'id'

feerrenrut committed:
Fix non-existent id attribute on DriverSetting instance.

dingpengyu commented:
Will pr currently merge 2019.3?

dingpengyu commented:
hi @leonardder 
This branch has conflicts that must be resolved
Conflicting files
source/visionEnhancementProviders/screenCurtain.py
thank

feerrenrut committed:
Update DriverSettingsMixin and derived

Class DriverSettingsMixin
- rename self._curDriverRef -> self._currentSettingsRef
- replace driver method with getSettings and _getSettingsStorage
  - getSettings to get settings definitions
  - _getSettingsStorage to load / save configuration values.
- renamed various internal only methods:
  - makeSliderSettingControl -> _makeSliderSettingControl
  - makeBooleanSettingControl -> _makeBooleanSettingControl
  - makeStringSettingControl -> _makeStringSettingControl
- settingsStorage is passed into the following:
  - _makeSliderSettingControl
  - _makeBooleanSettingControl
  - _makeStringSettingControl

Also update derived classes:
- VoiceSettingsPanel
- BrailleSettingsSubPanel
- VisionProviderSubPanel_Settings

feerrenrut committed:
Code cleanup

Fix linter errors / add documentation

feerrenrut committed:
Fix error on exit

feerrenrut committed:
Better docs in autoSettings class

feerrenrut committed:
Rename VisionEnhancementProviderStaticSettings

Static is misleading, this class contains the class
method _get_preInitSettings for settings that don't need an instance (or
to query an external dependency). It contains the instance method
_get_supportedSettings for settings that do need an instance (or to query
an external dependency).

feerrenrut committed:
Remove saveSettings param from terminate method on visionEnhancementProviderBase

feerrenrut committed:
Clarify naming, consolidate classes

Use getTranslatedName for AutoSettings classes
 - Driver
 - providerBase.VisionEnhancementProvider

Improve docs

feerrenrut committed:
Make screen curtain work

feerrenrut committed:
Use preInitSettings for NVDAHighlighter

There are no external dependencies change the settings for the highlighter.

feerrenrut committed:
Clarify abstract methods

Provide better docs about what derived classes must implement.

feerrenrut committed:
Prettify screen curtain and highlighter settings

feerrenrut committed:
Fix up toggle screen curtain command

Use 1 press for temp, and 2 presses for permanent.

With warning dialog:
 - enabling acts in "deferred mode"
 - dialog is refocused if already open and command is triggered again,
 the dialog is re-read in this case.
 - whether screen curtain is run in temp mode or not is updated in the background
Without warning dialog:
 - subsequent script calls repeat previous outcome. Otherwise three taps
  presses results in no speech

feerrenrut committed:
Add Example auto gui provider

This visionEnhancementProvider exercises the GUI auto construction for
vision provider panel.

feerrenrut committed:
Add visionEnhancementProvider Docs

feerrenrut committed:
Fix highlighter checkboxes being overwritten

feerrenrut committed:
Fix auto gui example

Differentiate between runtime settings and pre-runtime settings.

feerrenrut committed:
Rename autoGui.py

To make it clear this is just an example implementation.

feerrenrut committed:
Match name with filename

feerrenrut committed:
Demo different sources for GUI values

Values may come from:
- an external device/application.
- the config defaults.
- the last saved config.

feerrenrut committed:
Add copyright header and module doc.

feerrenrut committed:
Use dataclass for "provider info"

Rather than having to remember the order of these values, use names.
Currently the id of a provider must match it's module name.
Bundling these together (rather than just relying on them matching) allows
for more flexibility.

feerrenrut committed:
Use providerInfo rather than providerId: str to control providers

The providerInfo allows for differentiating providerId and moduleName.
This will be helpful when a provider wants a single implementation to
work with several versions of an app which have incompatible configurations.
In this case, they will need different config sections, hence different
providerId's.

feerrenrut committed:
Use a class to control start/stop of providers

This bundles together the several callables that were being passed around.

feerrenrut committed:
Improve docs

feerrenrut committed:
Provider ID no longer has to match module name

feerrenrut committed:
Fix linting errors

feerrenrut committed:
Remove unused role concept

feerrenrut committed:
Fix instance/class used in getSettingsPanelClass docs

feerrenrut committed:
Clarify return types of starting/stopping a provider

feerrenrut committed:
Give NVDA highlighter enable checkbox a better name and accelerator

feerrenrut committed:
Fix case error in name of internal function

feerrenrut committed:
Add copyright headers, and fix docs

feerrenrut committed:
Use single definition for SupportedSettingType

Since it is a iterable, len does not work directly. Convert to list first.

Commit:
Merge remote-tracking branch 'origin/master' into visionGui, fixing merge conflicts

feerrenrut committed:
Remove copied base __init__ implementation in Driver class

feerrenrut committed:
Comments for dialog announcement in toggle Screen curtain gesture

Commit:
Fix bad merge

Commit:
Fix lint error not caught by previous version of flake8tabs

feerrenrut committed:
fixup merge

feerrenrut committed:
Fix lint error

leonardder commented on source/visionEnhancementProviders/exampleProvider_autoGui.py:
```diff
@@ -0,0 +1,188 @@
+# A part of NonVisual Desktop Access (NVDA)
```
What are you intentions with this example? I assume it won't end up in the real pr? NOte that if you prefix the module name with an underscore, it will not be ignored by the framework.

feerrenrut replied:
> What are you intentions with this example?

At first I planned to just past it into the `readme.md` as example code. But I am worried it will become unmaintained, especially while this subsystem is new and prone to changes. At the moment I'm considering whether it can stay where it is and either return `False` from `canStart` or renamed to start with an underscore, if that stops it from being picked up. I want it to be easy to test with, since we don't have a built-in provider using the AutoGui panels.

>  it will not be ignored by the framework.

Do you mean it WILL be ignored?

leonardder commented on source/visionEnhancementProviders/screenCurtain.py:
```diff
@@ -32,19 +43,28 @@ class Magnification:
 	_MagShowSystemCursorFuncType = WINFUNCTYPE(BOOL, BOOL)
 	_MagShowSystemCursorArgTypes = ((1, "showCursor"),)
 
+	# initialise
```
```suggestion
	# initialize
```

leonardder commented on source/visionEnhancementProviders/screenCurtain.py:
```diff
@@ -66,30 +86,247 @@ class Magnification:
+		self.noButton.SetFocus()
+
+	def _onShowEvt(self, evt):
+		"""When no other dialogs have been opened first, focus lands in the wrong place (on the checkbox),
```
It's not clear to me when this happens. We're now setting focus to no both in _onDialogActivated and in this method. May be this one is obsolete now?

feerrenrut replied:
This is called when the dialog is "raised" or when focus returns to the dialog. In contrast, `_onDialogActivated` is called when the dialog is first created, on `show` or `showModal`

leonardder commented on source/visionEnhancementProviders/screenCurtain.py:
```diff
@@ -66,30 +86,247 @@ class Magnification:
+		if not settingsStorage.warnOnLoad:
+			return True
+		parent = self
+		dlg = WarnOnLoadDialog(
```
Could you put this into a with statement? This ensures that the dlg is destroyed when leaving the with statement.

feerrenrut replied:
I wondered about this, it probably doesn't hurt, and may reduce the chance of a bug, but is there a reason this must happen? Will it currently cause a bug I'm not thinking of?

leonardder replied:
I don't think it will hurt, it might take some time for the garbage collector to kick in and destroy the dialog.

Note that you can still return from a with statement

leonardder commented on user_docs/en/userGuide.t2t:
```diff
@@ -292,6 +293,7 @@ For example, if you are typing into an editable text field, the editable text fi
 
 The most common way of navigating around Windows with NVDA is to simply move the system focus using standard Windows keyboard commands, such as pressing tab and shift+tab to move forward and back between controls, pressing alt to get to the menu bar and then using the arrows to navigate menus, and using alt+tab to move between running applications.
 As you do this, NVDA will report information about the object with focus, such as its name, type, value, state, description, keyboard shortcut and positional information.
+When the [NVDA Highlighter #VisionNVDAHighlighter] is enabled, the location of the current system focus is also exposed visually.
```
See one of my previous comments about the name, which is now Focus Highlight in the gui.

leonardder commented on user_docs/en/userGuide.t2t:
```diff
@@ -1321,6 +1360,28 @@ This option won't be available if your braille display only supports automatic p
+The Vision category in the NVDA Settings dialog allows you to enable, disable and configure [vision enhancement providers #Vision].
+This settings category contains the following options:
+
+==== Vision enhancement providers ====[VisionSettingsVisionEnhancementProviders]
```
Ah, I hadn't realized that you didn't touch the userGuide yet. Do you want me to do this?

feerrenrut replied:
I think we should get the rest of the interface sorted first. Then we can circle back to this.

LeonarddeR reviewed (commented):
I really, really like this approach and think it is very solid! Thanks for your patience and efforts to make this like it is now!

feerrenrut commented on source/driverHandler.py:
```diff
@@ -163,7 +91,7 @@ def _paramToPercent(cls, current, min, max):
+		return paramToPercent(current, min, max)
 
 	@classmethod
 	def _percentToParam(cls, percent, min, max):
```
I kept these because I didn't want to change the interface of the `driverHandler` module. Even these methods are preceded by an underscore, and should be considered private, I can't verify that no external driver is using them. There is also very little cost to keeping them around.

leonardder replied:
But, as driverHandler.Driver inherrits from autoSettingsUtil.autoSettings.AutoSettings, it will already have these methods from there. You're not breaking anything when removing it. Or am I just missing something essential here?

feerrenrut replied:
Yes, you are right. I'll move the documentation too.

feerrenrut commented on source/gui/settingsDialogs.py:
```diff
@@ -1030,169 +1044,248 @@ def __call__(self,evt):
 
-class DriverSettingsMixin(object):
+
+class DriverSettingsMixin(metaclass=ABCMeta):
```
I didn't want to do this to be cautious of backwards compatibility. I might rename it and add an alias with a comment saying it is deprecated.

leonardder replied:
How it stands now, i don't think we should take backwards compatibility with the driver system into account. It was introduced as such in NVDA 2019.2 and therefore is only one release old. I doubt whether it has ever been used outside of core. Of course external drivers that have been ported to Python 3 might be using these classes now, but 2019.3 is still not set in stone and subject to change.

I leave the final decision up to you, though.

feerrenrut committed:
Use type hinting rather than docstring

feerrenrut committed:
Rename DriverSettingsMixin to AutoSettingsMixin

Provides better consistency with the AutoSettings class.
An alias mapping DriverSettingsMixin to AutoSettingsMixin is provided for
backwards compatibility, however DriverSettingsMixin should be considered
deprecated. Use AutoSettingsMixin instead.

feerrenrut commented on source/gui/settingsDialogs.py:
```diff
@@ -1030,169 +1044,248 @@ def __call__(self,evt):
 		self.lastControl=checkbox
 		return checkbox
 
 	def updateDriverSettings(self, changedSetting=None):
```
Again, I want to be cautious of maintaining compatibility. If I was to rename this, I would probably call it `UpdateGuiForSettings` or similar.

feerrenrut commented on source/gui/settingsDialogs.py:
```diff
@@ -1030,169 +1044,248 @@ def __call__(self,evt):
+				if isinstance(setting, NumericDriverSetting):
+					settingMaker = self._makeSliderSettingControl
+				elif isinstance(setting, BooleanDriverSetting):
+					log.debug(f"creating a new bool driver setting: {setting.id}")
```
I was having trouble getting things working, and was only using a `BooleanDriverSetting`. I think this should be removed.

feerrenrut committed:
Add requirements doc string for classes using the AutoSettingsMixin

feerrenrut committed:
Remove debug logging from AutoSettingsMixin

feerrenrut commented on source/gui/settingsDialogs.py:
```diff
@@ -2879,6 +3051,332 @@ def onBlinkCursorChange(self, evt):
+					self._parent
+				)
+
+	def getProviderInstance(self) -> Optional[vision.providerBase.VisionEnhancementProvider]:
```
Ah yes, sorry. This needs more work.

feerrenrut commented on source/vision/visionHandler.py:
```diff
@@ -48,18 +49,73 @@ def getProviderClass(
+
+def getProviderInfo(providerId: providerInfo.ProviderIdT) -> Optional[providerInfo.ProviderInfo]:
+	# This mechanism of getting the provider list and looking it up is particularly inefficient, but, before
+	# refactoring, confirm that getProviderList is / isn't cached.
```
I think this approach is fine for now. But it might be worth investigating whether it can be made more efficient. I don't really like the idea of generating this list so often. One benefit though is that it would respond to newly added files. If the list is cached by the autoProperty code, then it might not require any work.

feerrenrut commented on source/vision/visionHandler.py:
```diff
@@ -71,33 +127,43 @@ def postGuiInit(self) -> None:
 		config.post_configProfileSwitch.register(self.handleConfigProfileSwitch)
 
-	def terminateProvider(self, providerName: str) -> bool:
+	def terminateProvider(
```
Hmm. I was trying to keep a consistent interface. Rather than using `providerId` as the token for a provider, using `ProviderInfo`. Another option which might get around a few of these issues is to have the `VisionHandler` class maintain a dictionary of `providerId` to `ProviderInfo` objects. This way, when it needs to know the associated module name etc it is easy to look up. We would then revert these methods to take `providerId`, perhaps add one to `getProviderInfo` from a `providerId`. Thoughts?

leonardder replied:
I think that makes sense.

feerrenrut replied:
No, looking at this again, I'd prefer to keep this functions using `providerId`. All the places that use this already have access to and need the full `providerInfo`. I think this approach is more extensible, if these methods need some other information about a provider it is easy to include. For the python console I think it reasonable to do:
```
import vision
vision.handler.terminateProvider(vision.visionHandler.getProviderInfo("screenCurtain"))
```
Though I think I will move a this around slightly so that it is instead:
```
import vision
vision.handler.terminateProvider(vision.handler.getProviderInfo("screenCurtain"))
```

leonardder replied:
Sounds good.

feerrenrut commented on source/visionEnhancementProviders/NVDAHighlighter.py:
```diff
@@ -184,50 +186,179 @@ def refresh(self):
+	@classmethod
+	def getTranslatedName(cls) -> str:
+		# Translators: Description for NVDA's built-in screen highlighter.
+		return _("Focus Highlight")
```
I think it is worth considering end users and developers can have different interpretations of "focus". While the technical definition is certainly not very accurate here, for someone looking at the screen, their "focus" is where they currently are (or should be) looking.

leonardder replied:
True. I think NVDA's use of the word focus in the GUI mostly refers to the real system focus, see the review cursor settings, for example.

feerrenrut replied:
Yeah, we should be cautious about overloading the term. It might be nice to see what a few low vision users think?

feerrenrut commented on source/visionEnhancementProviders/NVDAHighlighter.py:
```diff
@@ -249,20 +381,20 @@ def __init__(self):
 			self._highlighterThread = None
 		winGDI.gdiPlusTerminate()
 		self.contextToRectMap.clear()
-		super(VisionEnhancementProvider, self).terminate()
```
Perhaps it should, but note that `VisionEnhancementProvider.terminate` is abstract and not implemented.

feerrenrut committed:
Fix some issues in docs

feerrenrut committed:
Fix multiple provider error handling.

One message is now shown.
Code is shared better between starting single providers and
starting multiple providers.

feerrenrut committed:
Ensure destruction of warning dialog for screencurtain

feerrenrut committed:
Improve readability of make settings method.

leonardder commented on source/vision/providerBase.py:
```diff
@@ -7,35 +7,131 @@
-		"""Reinitializes a vision enhancement provider, reusing the same instance.
+
+	@classmethod
+	def getSettings(cls) -> VisionEnhancementProviderSettings:
```
Do we want to prescripe that a provider has settings? IN that case, I think this method should be abstract:

```suggestion
	@abstractmethod
		def getSettings(cls) -> VisionEnhancementProviderSettings:
```

feerrenrut commented on source/gui/settingsDialogs.py:
```diff
@@ -2879,6 +3051,332 @@ def onBlinkCursorChange(self, evt):
+			# Terminating a provider from the gui should never save the settings.
+			# This is because termination happens on the fly when unchecking check boxes.
+			# Saving settings would be harmful if a user opens the vision panel,
+			# then changes some settings and disables the provider.
```
Can the saveSettings parameter be removed entirely? Settings should only be saved when the rest
of the configuration is saved. The extension points for config can be used for that.
Perhaps there is a case I am not thinking of?

leonardder replied:
If we're terminating a provider by hand, we should make sure that the settings are saved. I don't think that happens if we remove the parameter, right?

feerrenrut replied:
Yes, you are right. I think the confusion here for me was thinking that this would save the config to the settings file. I wonder if we could rename this to make it clearer; `updateConfigSettings` perhaps?

feerrenrut committed:
Fix ScreenCurtain enable checkbox state after initialisation error

feerrenrut committed:
Fix initial state of screenCurtain enabled checkbox

feerrenrut committed:
Refactor updateDriverSettings in in AutoSettingsMixin

Extract implementation details into helper methods to highlight high level
logic.

feerrenrut commented on source/gui/settingsDialogs.py:
```diff
@@ -2879,6 +3051,332 @@ def onBlinkCursorChange(self, evt):
+				settingsPanel = settingsPanelCls(parent=self, providerControl=providerControl)
+			# E722: bare except used since we can not know what exceptions a provider might throw.
+			# We should be able to continue despite a buggy provider.
+			except:  # noqa: E722
```
This is still much broader than I would like in normal code. The comment is still true, so I will leave something similar here.

feerrenrut committed:
Remove noqa comment for bare Except

Instead catch Exception.

feerrenrut committed:
Change error message for a provider GUI that can not be created

feerrenrut committed:
Simplify _enableToggle logic

feerrenrut committed:
Simplify and reorder imports

feerrenrut committed:
Remove package level getProviderList method

feerrenrut committed:
Fix typo in comment

feerrenrut commented on source/vision/visionHandler.py:
```diff
@@ -108,58 +174,50 @@ def terminateProvider(self, providerName: str) -> bool:
 		"""
 		Enables and activates the supplied provider.
-		@param providerName: The name of the registered provider.
+		@param provider: The provider to initialise.
```
Oh haha. British / Australian spelling vs American. Though I agree, American spelling is more prevalent in code, we should try to be consistent with that.

leonardder replied:
Wow, I hadn't realised that this was valid.

feerrenrut committed:
Clarify comment

feerrenrut committed:
Use US spelling for initialize

feerrenrut committed:
Clarify comments on getSettingsPanelClass(cls)

feerrenrut committed:
Fix spelling of NVDAHighlighter class

feerrenrut committed:
Fix unused imports

feerrenrut committed:
Remove duplicated methods

Driver inherits from AutoSettings, which already provides these methods.
Moved the documentation to the AutoSettings impl.

feerrenrut committed:
Fix missing call to super from terminate

dingpengyu commented:
hi leonardder
Currently pr can't be built normally, I want to get the latest test build to test, can fix the appveyor error?
thanks

feerrenrut committed:
Change getSettings to abstract

Implementors of VisionEnhancementProvider must provide a settings instance.

feerrenrut commented on source/vision/visionHandler.py:
```diff
@@ -108,58 +174,50 @@ def terminateProvider(self, providerName: str) -> bool:
-				# therefore it is unknown what to expect.
-				log.error("Error while reinitializing provider %s" % providerName, exc_info=True)
-				return False
+			providerInst.reinitialize()
```
I'll wrap it in a try block to make it clear. But I will re-raise the exception since this is the control flow we want anyway. Note that this is only called from places that should handle exceptions anyway.

feerrenrut committed:
Catch any error on re-init

terminate method inside providerInst.reinitialize may raise. Even though
all callers to VisionHandler.initializeProvider should handle exceptions
anyway, adding a try block to log it and re-raise clarifies the control
flow.

feerrenrut committed:
Fix lint

feerrenrut committed:
Fix translation test

feerrenrut committed:
Rename example provider

So that it no longer appears in NVDA settings.

feerrenrut committed:
Add instructions to test the example provider

dingpengyu commented:
thanks 
Appveyor error has been resolved

leonardder commented on source/gui/settingsDialogs.py:
```diff
@@ -268,31 +275,35 @@ class SettingsPanel(wx.Panel, DpiScalingHelperMixin, metaclass=guiHelper.SIPABCM
 
 	@abstractmethod
-	def makeSettings(self, sizer):
+	def makeSettings(self, sizer) -> wx.BoxSizer:
```
I think this annotation is incorrect.

```suggestion
	def makeSettings(self, sizer: wx.BoxSizer) -> None:
```

leonardder commented on source/gui/settingsDialogs.py:
```diff
@@ -2663,6 +2665,9 @@ def onOk(self, evt):
+			# Translators: This message is presented when
+			# NVDA is unable to load the selected
+			# braille display.
 			gui.messageBox(_("Could not load the %s display.")%display, _("Braille Display Error"), wx.OK|wx.ICON_WARNING, self)
```
I agree

leonardder commented on source/autoSettingsUtils/utils.py:
```diff
@@ -0,0 +1,58 @@
+"""
+
+
+def paramToPercent(current, min, max):
```
```suggestion
def paramToPercent(current: int, min: int, max: int) -> int:
```

leonardder commented on source/autoSettingsUtils/utils.py:
```diff
@@ -0,0 +1,58 @@
+	@param max: The maximum value.
+	@type max: int
+	"""
+	return int(round(float(current - min) / (max - min) * 100))
```
In python 3, the behavior of round was changed to return an integer if the number of digits is omitted.

```suggestion
	return round((current - min) / (max - min) * 100)
```

leonardder commented on source/autoSettingsUtils/utils.py:
```diff
@@ -0,0 +1,58 @@
+	return int(round(float(current - min) / (max - min) * 100))
+
+
+def percentToParam(percent, min, max):
```
```suggestion
def percentToParam(percent: int, min: int, max: int) -> int:
```

leonardder commented on source/autoSettingsUtils/utils.py:
```diff
@@ -0,0 +1,58 @@
+	@param max: The maximum raw parameter value.
+	@type max: int
+	"""
+	return int(round(float(percent) / 100 * (max - min) + min))
```
```suggestion
	return round(percent / 100 * (max - min) + min)
```

leonardder commented on source/autoSettingsUtils/utils.py:
```diff
@@ -0,0 +1,58 @@
+		"""
+		self.id = id
+		self.displayName = displayName
+		# Keep backwards compatibility
```
I thought I already pointed out that these can be removed, but that might have not been posted correctly.

leonardder commented on source/gui/settingsDialogs.py:
```diff
@@ -268,31 +275,35 @@ class SettingsPanel(wx.Panel, DpiScalingHelperMixin, metaclass=guiHelper.SIPABCM
 		self.settingsSizer=wx.BoxSizer(wx.VERTICAL)
 		self.makeSettings(self.settingsSizer)
-		self.mainSizer.Add(self.settingsSizer, flag=wx.ALL)
+		self.mainSizer.Add(self.settingsSizer, flag=wx.ALL | wx.EXPAND)
```
What does this addition of wx.EXPAND do?

feerrenrut replied:
The `wx.EXPAND` flag tells wx to allow this control to take up space opposite to the BoxSizers orientation. So in this case, it means that the `settingsSizer` can grow horizontally as room becomes available in the `mainSizer`

leonardder commented on source/vision/providerBase.py:
```diff
@@ -7,35 +7,130 @@
 
-class VisionEnhancementProvider(driverHandler.Driver):
-	"""A class for vision enhancement providers.
+class VisionEnhancementProviderSettings(AutoSettings, ABC):
```
I believe that AutoSettings is already using ABC, as it inherrits from AutoPropertyObject.

```suggestion
class VisionEnhancementProviderSettings(AutoSettings):
```

leonardder commented on source/vision/providerBase.py:
```diff
@@ -7,35 +7,130 @@
 	@abstractmethod
-	def registerEventExtensionPoints(self, extensionPoints: EventExtensionPoints):
+	def terminate(self) -> None:
+		"""Terminate this driver.
```
```suggestion
		"""Terminate this provider.
```

leonardder commented on source/vision/providerBase.py:
```diff
@@ -7,35 +7,130 @@
+	- AutoSettings.getTranslatedName:
+			The string that should appear in the GUI as the name
+	- AutoSettings._get_supportedSettings:
+			The "runtime" settings for your provider
```
I think that the vision framework introduces some confusion here.

1. Drivers inherrit from AutoSettings. A driver instance is the same as an auto settings instance. Therefore, for AutoSettings, this property returns the settings supported at runtime of the driver.
2. For vision enhancement providers however, an instance of this class is saved on the provider's class. This means that supportedSettings is used to retrieve the settings for the provider, even **before** the provider is actually initialized.

I think this should be clear from the doc string of this class. Therefore, I think the doc string should state that:

* Every vision enhancement provider that implements settings holds a reference to an instance of this class, not the actual class itself.
* For settings that are strictly runtime, the vision enhancement provider should instruct the settings instance at initialization time that it should expose the runtime only settings using this property. exampleProvider_autoGui is a perfect example of this.

feerrenrut replied:
> This means that `supportedSettings` is used to retrieve the settings for the provider, even before the provider is actually initialized.

Yes, my understanding of the `AutoSettings` method `_get_preInitSettings` is that, as a class method, it relates to the pre-init of the `AutoSettings` class itself, rather than the driver / provider class it might be associated with. Since we have an instance of the settings class, we should use `_get_supportedSettings`. One benefit with this approach is to decouple the settings behaviour from the lifetime of the provider. 

Alternatively, we could choose to communicate the state of the provider to the settings class by calling `_get_preInitSettings` before the provider is initialised, and `_get_supportedSettings`, though I am a bit worried this isn't very explicit.

Another alternative is that we return a class type from `getSettings` on the provider, and we initialise it at the same time as the provider. Though I this adds extra complexity and doesn't really get us anything. 

Also, I note that `_get_preInitSettings` isn't really used anywhere in the code base. We could consider removing it from `AutoSettings`?  Then, in the future, if we wish to update synth / braille drivers to support pre-init settings we can look at what we learn from vision providers.

leonardder replied:
Hmm, I actually think that the code is fine as it is. I think just some extra comments being more explicit about when applies what will solve my concerns.

leonardder commented on source/visionEnhancementProviders/_exampleProvider_autoGui.py:
```diff
@@ -0,0 +1,190 @@
+			options.
+		"""
+		settings = self.getSettings()
+		settings.addRuntimeSettingsAvailibility([
```
I think it would help if this could be abstracted a bit, so that it is easier for custom providers to add runtime only settings without copying much of this code.

How about the following:

1. On VisionEnhancementProviderSettings, add a registerInstance method. When called, the method should store a weakref to the instance of the provider on the instance of VisionEnhancementProviderSettings.
1. The base implementation of VisionEnhancementProvider can call this registerInstance method when initializing. It can optionally unregister itself when calling terminate, though a weakref will deal with unregistration automatically.
1. VisionEnhancementProviderSettings can expose the runtime only settings when getting the supportedSettings property and there is a registered instance. If not, it will only expose the preInit settings.

feerrenrut replied:
I think I'd prefer not to be prescriptive about this, since we don't have any current inbuilt code depending on runtime settings. The `_exampleProvider_autoGui.py` shows one way you might do this. 

If you are happy to try to get this merged as it is, perhaps you can make a draft PR afterwards to explore these ideas further?

leonardder replied:
Yeah, I"m fine with that, though filing the draft probably won't have that much priority. If there's one or even more ways to do it, that's fine, as long as it is possible.

leonardder commented on source/gui/settingsDialogs.py:
```diff
@@ -2879,6 +3074,392 @@ def onBlinkCursorChange(self, evt):
+
+	def _enableToggle(self, evt):
+		if evt.IsChecked():
+			self._providerControl.startProvider()
```
If this fails, the check box still remains checked. Therefore, if enabling the check box results in an error, it should be disabled again and we should refrain from updating the gui.

feerrenrut replied:
Good catch

feerrenrut committed:
Clean up param/percent conv functions

Restore type hints, remove types from doc-string.
Remove unncessary cast

feerrenrut committed:
Add missing translator comments

feerrenrut committed:
remove unused properties

feerrenrut committed:
Fix incorrect type hint

feerrenrut committed:
Fix doc-string

feerrenrut committed:
Use ABC from base class

feerrenrut committed:
Refactor provider list generation

cache provider lists.

feerrenrut committed:
Make getProviderClass private

Instead it is exposed through ProviderInfo class

feerrenrut committed:
Move getProviderList and getProviderInfo functions

Make them instance methods on VisionHandler for greater API consistency.
There is also very little sense in calling these before the handler is initialised.

feerrenrut committed:
Fix no contextmanager

feerrenrut committed:
Fix documentation for runtime settings

feerrenrut committed:
Fix error on termination handling for autoGui providers

feerrenrut committed:
Fix parent windows binding to AutoSettingsMixin controls

This matches how checkboxes are handled.

Commit:
AutoSettings.percentToParam should be private, with an underscore

feerrenrut committed:
Make VisionHandler.providers private

Provide alternernative convenience functions to cater to the uses of the
providers property. This allows us to more easily change the structures
used for storing active providers / providerInfo / and their mapping with
providerId.

dingpengyu commented:
hi This branch has conflicts that must be resolved
Conflicting files
source/globalCommands.py
thanks

LeonarddeR reviewed (commented):
> hi This branch has conflicts that must be resolved

This is a known issue. This work is a continuous work in progress.

Commit:
Update user guide

Commit:
Fix square brackets

Commit:
User docs: visual aide> visual aid

feerrenrut committed:
Fix spelling: aides vs aids

leonardder commented on source/gui/settingsDialogs.py:
```diff
@@ -2879,6 +3083,387 @@ def onBlinkCursorChange(self, evt):
+			return
+		# Able to successfully start / terminate:
+		self._providerSettings.updateDriverSettings()
+		self._providerSettings.onPanelActivated()
```
I think we need to skip the event here

```suggestion
		self._providerSettings.onPanelActivated()
		evt.Skip()
```

feerrenrut replied:
It probably wouldn't hurt, but I don't see why it would be necessary. Skip is only useful if there are parent windows that might want to catch the event. I can't think of a situation where this would happen for the wrapper.

leonardder replied:
Agreed.

feerrenrut committed:
Handle providers with no options with auto gui

feerrenrut commented on source/gui/settingsDialogs.py:
```diff
@@ -1018,6 +1018,7 @@ def driver(self):
 		return self._driverRef()
 
 	def __call__(self,evt):
+		evt.Skip()  # allow other handlers to also process this event.
```
I had a look at the classes that use AutoSettings mixin (`BrailleSettingsSubPanel`, and `VoiceSettingsPanel`). Only braille calls `bind` at all, and it does so on explicitly on checkboxes that it has created. It has no "catch-all" checkbox event handler, which is a good thing.

feerrenrut committed:
Remove saveSettings param from Driver.terminate


It was introduced with this feature, but is not used.

feerrenrut committed:
Remove preInitSettings property

This was introduced with this feature.
Not being used by the framework, and seems likely to mislead.

feerrenrut committed:
Rename getTranalatedName -> getDisplayName

- class AutoSettings
- data class ProviderInfo

This is more consistent with class DriverSettings
It communicates the intended usage of the string better.

feerrenrut committed:
Fix consistency of imports.

LeonarddeR reviewed (commented):
This looks awesome, thanks @feerrenrut

feerrenrut commented:
This work was delivered via commit 46c5606c27682d4c37905f93eb0eac11149123f3

nishimotz commented:
Thank you for beautifully reimplementing my work.
I would like to ask from the translation standpoint.
Is the Browse Mode Cursor and the Review Cursor technically different?

ruifontes commented:
Browse cursor is the one used in browse mode in the web navigators, PDF 
files and so on...

Review cursor is used on object navigation and screen review...


Rui Fontes


Às 06:30 de 24/11/2019, Takuya Nishimoto escreveu:
>
> Thank you for beautifully reimplementing my work.
> I would like to ask from the translation standpoint.
> Is the Browse Mode Cursor and the Review Cursor technically different?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub 
> <https://github.com/nvaccess/nvda/pull/10082?email_source=notifications&email_token=ADZAPRU5BLY4X4D5ART7M5LQVINPTA5CNFSM4ILJQPCKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEFAEYWA#issuecomment-557861976>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/ADZAPRTMFQIUPNP7Y7FPXDTQVINPTANCNFSM4ILJQPCA>.
>

nishimotz commented:
Thank you for the suggestion.
Just confirmed that "Highlight browse mode cursor" does not work with review cursor.

STR:
* Open NVDA menu and select Configuration Profile dialog.
* Press NVDA+right arrow key (laptop layout)
* The move of review cursor can be confirmed. However, the yellow rectangle is not available.

I understand the description regarding the vision framework is correct.

So far, "browse mode cursor" is used only in the Content Recognition chapter of the User Guide and almost undocumented.

leonardder commented:
The browse mode caret is the virtual caret used in all browse mode 
situations, so also in browsers. It isn't the same as the review cursor 
and therefore indeed, the review cursor position isn't shown. Most 
important reason of this is that we don't always have the physical 
position of the review cursor, so we can't reliably show it.
